File: 113-tree-path-sum-ii.java
/*
  Source : https://leetcode.com/problems/path-sum-ii/
  Author : Alex Xu
  Date   : Mar 16, 2015
 
  Problem:
  Given a binary tree and a sum, find all root-to-leaf paths where each path's sum equals the given sum.

  For example:
  Given the below binary tree and sum = 22,
   5
  / \
  4   8
  /   / \
  11  13  4
  /  \    / \
  7    2  5   1
  return
  [
  [5,4,11,2],
  [5,8,4,5]
  ]
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {      
        if (root == null) {
            List<List<Integer>> paths = new LinkedList<List<Integer>>();    
            return paths;
        }
        
        if (root.left == null && root.right == null && root.val == sum) {
            // found a path, then create a path and return it
            List<Integer> path = new LinkedList<Integer>();
            path.add(root.val);
            List<List<Integer>> paths = new LinkedList<List<Integer>>();            
            paths.add(path);
            
            return paths;
        }
        
        List<List<Integer>> leftPaths = pathSum(root.left, sum - root.val);
        List<List<Integer>> rightPaths = pathSum(root.right, sum - root.val);
        leftPaths.addAll(rightPaths);
        
        for (List<Integer> path : leftPaths) {
            path.add(0, root.val);
        }
        
        return leftPaths;
    }
}

/**
 * 
 *
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> paths = new LinkedList<List<Integer>>();  
        List<Integer> path = new LinkedList<Integer>();
        dfs(root, sum, paths, path);
        return paths;    
    }
    
    public void dfs(TreeNode root, int sum, List<List<Integer>> paths, List<Integer> path) {
        if (root == null)
            return;
        
        path.add(root.val);
        if (root.left == null && root.right == null) {
            if (sum == root.val) {
                // found a path, then add it             
                paths.add(path);
            }
        } else {
            List<Integer> copyPath = new LinkedList<Integer>();
            copyPath.addAll(path);
   
            // find path from left
            dfs(root.left, sum - root.val, paths, path);
            // find path from right
            dfs(root.right, sum - root.val, paths, copyPath);                
        }
    }
}

File: 113-tree-path-sum-ii.py
"""
  Source : https://leetcode.com/problems/path-sum-ii/
  Author : Alex Xu
  Date   : Mar 16, 2015
 
  Problem:
  Given a binary tree and a sum, find all root-to-leaf paths where each path's sum equals the given sum.

  For example:
  Given the below binary tree and sum = 22,
   5
  / \
  4   8
  /   / \
  11  13  4
  /  \    / \
  7    2  5   1
  return
  [
  [5,4,11,2],
  [5,8,4,5]
  ]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        paths = []
        
        if not root:
            return paths
        
        if root.left is None and root.right is None:
            if sum == root.val:
                # found one path, create the path with leaf node's value
                path = [root.val]
                paths.insert(0, path)
            return paths
        
        # search from left and right sub tree
        paths = self.pathSum(root.left, sum - root.val) + self.pathSum(root.right, sum - root.val)
        if paths:
            for path in paths:
                path.insert(0, root.val)
                
        return paths
